include/ttmlir/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.h

```diff
@@ -0,0 +1,21 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_CONVERSION_TTIRTOTTIR_TTIRTOTTIR_H
+#define TTMLIR_CONVERSION_TTIRTOTTIR_TTIRTOTTIR_H
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir::tt {
+
+void populateTTIRToTTIRDecompositionPatterns(MLIRContext *ctx,
+                                             RewritePatternSet &patterns,
+                                             TypeConverter &typeConverter);
+
+std::unique_ptr<OperationPass<ModuleOp>> createTTIRToTTIRDecompositionPass();
```

Comment:
I think we shouldn't call this a "decomposition" pass, as we'll potentially fuse ops as well. Can we call it `WideTTIRtoNarrowTTIR`?

@nsmithtt I was talking to @rpavlovicTT and @sdjordjevicTT about this. Is the idea to have TTIRtoTTNN pass be completely mechanical? I.e. have wide-to-narrow-ttir-pass consume all the complexity of mapping 3rd party TTIR ops to ttnn-consumable TTIR ops?

Reply:
Before it starts to get messy, I think we should develop a high level design for these 3 steps:

1) Legalization

The above decomposition pass is an example of legalization i.e. at this level all ops should be converted to legal TTIR ops. This stage is our primary target for this sprint where we want to legalize the ops from all front-ends.

2) Canonicalization

As @sdjordjevicTT mentioned above, this is purely an optimization stage. Even if we don't plan to do much optimization in this sprint, but I think we should still have infrastructure in place to run canonicalization or optimization passes.

I think both Legalization and Canonicalization should become part of TTIRToTTIR pipeline. 

3) Lowering

At this stage we perform TTIRToTTNN or TTIRToTTMetal conversions and should be specific to each backend.

Reply:
> Can we call it WideTTIRtoNarrowTTIR?

If we're not going with decomposition I'd prefer to call it like `TTIRToPrimTTIR` or something.

> @nsmithtt I was talking to @rpavlovicTT and @sdjordjevicTT about this. Is the idea to have TTIRtoTTNN pass be completely mechanical? I.e. have wide-to-narrow-ttir-pass consume all the complexity of mapping 3rd party TTIR ops to ttnn-consumable TTIR ops?

I'm not sure it has to be so black and white, I think the idea is to mimic what we did in Buda and progressively lower.  This TTIR decomp pass should try its best to lower to some set that better maps to TTNN such that the TTNN dialect conversion could be as mechanical as possible, but it might not be perfect and I don't think we should bring new ops into TTIR just to have better mappings to TTNN.

Reply:
>If we're not going with decomposition I'd prefer to call it like TTIRToPrimTTIR or something.

That's fine by me, I just don't want it to be called decomp if it'll support fusing as well.

>I'm not sure it has to be so black and white, I think the idea is to mimic what we did in Buda and progressively lower. This TTIR decomp pass should try its best to lower to some set that better maps to TTNN such that the TTNN dialect conversion could be as mechanical as possible, but it might not be perfect and I don't think we should bring new ops into TTIR just to have better mappings to TTNN.

Awesome, thanks for clearing it up!

Reply:
My take here would be to separate clearly responsibilities between decomposition and fusion. We should implement fusion as a separate standalone pass. Separating it from decomposition provides clear modularity and flexibility, making it easier to maintain and tune each pass individually. This keeps its functionality distinct, ensuring that fusion optimizations are applied where appropriate without interfering with decomposition's role of simplifying wide operations.

Probably we want to run fusion after decomposition, as the decomposition pass will have already broken down wide operations into narrow TTIR ops, making fusion easier to apply to the resulting primitives.

Canonicalization can assist by simplifying patterns or normalizing forms, it's best not to rely on it as the sole mechanism for fusion. Instead, we should keep it as a supporting pass that prepares ops for fusion and other optimization passes. This way, fusion and optimizations can leverage a well-prepped intermediate representation.

Reply:
I'll just add this last piece as I feel I wasn't understood, but won't comment anymore and will leave it to you guys.

Are we 100% sure we'll be able to do "wide-to-narrow" without fusing anything? There are already some reduction ops that @mrakitaTT encountered, which are done via several ops in StableHLO, but can be "lowered" into one TTIR/TTNN op. Calling this pass a decomposition pass means introducing an assumption that doesn't need to be introduced today.

Reply:
We will have fusing as a separate pass, I'm still not sure why you think naming is a problem here.

Reply:
Yeah, I would fuse them in a separate pass and clearly separate responsibilities between them.